### PR TITLE
JIT: Factor loop duplication code

### PR DESCRIPTION
**Pull request by jakobbotsch**
Factor the loop duplication code out of loop cloning and loop unrolling in anticipation of also using it in loop peeling.

No diffs expected.

### CODE REVIEW

File: src/coreclr/jit/flowgraph.cpp [hunk 1]
```diff
@@ -5409,6 +5409,176 @@ bool FlowGraphNaturalLoop::HasDef(unsigned lclNum)
...
+//   bottomNeedsRedirection - Whether or not to insert a redirection block for the bottom block in case of fallthrough
+//
+// Remarks:
+//   Due to fallthrough this block may need to insert blocks with no
+//   corresponding source block in "map".
+//
+void FlowGraphNaturalLoop::Duplicate(BasicBlock**     insertAfter,
+                                     BlockToBlockMap* map,
+                                     weight_t         weightScale,
+                                     bool             bottomNeedsRedirection)
```
Comment (jakobbotsch) [inline src/coreclr/jit/flowgraph.cpp:5464]: FWIW, we will be able to get rid of this `bottomNeedsRedirection` parameter once we don't have fallthrough anymore. We could've probably gotten rid of it here with some complications, but I think it is going to be much easier once we don't have fallthrough, so I didn't want to bother.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @BruceForstall 

No diffs.

File: src/coreclr/jit/flowgraph.cpp (hunk 1 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/flowgraph.cpp:5464]: Thanks for pointing this out. This PR will probably get merged before #97488, so I'll plan on removing this over there.

**Review by BruceForstall — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Hmm, there is actually a single diff in linux-arm64. Going to check what that is, but I expect it to just be something around block weighting that is subtly different...

**Comment by jakobbotsch** [comment]
Ah, the diff is because the PR removes this quirk in unrolling:
https://github.com/dotnet/runtime/blob/c105e7da368ad23097e20e51b0ad054b771dab7c/src/coreclr/jit/optimizer.cpp#L1671-L1674

Didn't realize it was still there, but given the single diff that doesn't seem necessary to do separately.

**Pull request merged**